### PR TITLE
Added handling for the inline view container elements in the `text/plain` data transfer

### PR DESCRIPTION
**Pull request by Dumluregn**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (clipboard): Added handling for the elements with `dataPipeline:transparentRendering` property in the `text/plain` data transfer. Closes https://github.com/ckeditor/ckeditor5/issues/16583.

---

### Additional information

Theoretically, we could also solve it without adding a custom element property, but it's more complex: we need to convert the element to model one and then check in schema if it is an inline object or something similar.

### CODE REVIEW

**Review by scofalik — CHANGES_REQUESTED** [review]
I don't like this magic property -- it's very easy to forget about setting such thing. If we go this way, we should probably add a method `view.Writer#createInlineContainerElement()` and then change our implementations if we have container inline elements anywhere (probably not?).

Instead, I see two different ways we can go about it:

1.  I think I earlier raised my doubt whether merge field should be a container element. It's more like an empty container, i.e. it is atomic, indivisible and you don't "edit inside it". AFAIK `EmptyElement` cannot be used from technical reasons, so I propose `RawElement`. If not that, then it is a bit problematic.
2.  Instead of introducing a new entity ("inline container element") and a new property for it, we can base on `dataPipeline:transparentRendering`. Such view element should be transparent when converting view to final data output. So, if it should be transparent, it should not add any padding (although its children can - I am not sure how exactly this method is used, whether it is fired recursively, I hope so).

I like both solutions, but I think that 2nd is also a good generic solution, so if we want to postpone discussion about `RawElement`, we can go with 2nd solution for now and it will not be incorrect even if we swap to `RawElement` anyway.

**Comment by niegowski** [comment]
> I don't like this magic property -- it's very easy to forget about setting such thing. If we go this way, we should probably add a method `view.Writer#createInlineContainerElement()` and then change our implementations if we have container inline elements anywhere (probably not?).

I'd suggest using `DomConverter#blockElements` property to verify where we should separate with newlines.

**Comment by scofalik** [comment]
That's a good suggestion too. We are talking about `<span>`s here, and they should not be marked as `blockElements`, so it will work too.

Although still I think that ignoring transparent elements makes sense, both improvements will fix our problem and make sense on their own.

**Comment by niegowski** [comment]
> inline image is `` which I guess is block?

Those are `inlineObjectElements`.

**Comment by Dumluregn** [comment]
I went with the `dataPipeline:transparentRendering` solution after all.

**Comment by scofalik** [comment]
LGTM. @niegowski is it okay for you?

File: packages/ckeditor5-clipboard/src/utils/viewtoplaintext.ts [hunk 1]
```diff
@@ -86,6 +86,12 @@ function newLinePadding(
 		return '\n';
 	}
 
+	// Do not add padding around the elements that won't be rendered.
+	if ( ( element.is( 'element' ) && element.getCustomProperty( 'dataPipeline:transparentRendering' ) ) ||
+		( previous.is( 'element' ) && previous.getCustomProperty( 'dataPipeline:transparentRendering' ) ) ) {
```
Comment (niegowski) [inline packages/ckeditor5-clipboard/src/utils/viewtoplaintext.ts:91]:
```suggestion
	if ( 
		element.is( 'element' ) && element.getCustomProperty( 'dataPipeline:transparentRendering' ) ||
		previous.is( 'element' ) && previous.getCustomProperty( 'dataPipeline:transparentRendering' )
	) {
```

**Review by niegowski — APPROVED** [review]

**Comment by scofalik** [comment]
Thanks @niegowski, we will wait with merge until merge fields feature is ready.

**Review by scofalik — APPROVED** [review]

File: packages/ckeditor5-clipboard/src/utils/viewtoplaintext.ts [hunk 2]
```diff
@@ -86,6 +86,14 @@ function newLinePadding(
 		return '\n';
 	}
 
+	// Do not add padding around the elements that won't be rendered.
+	if ( 
+		element.is( 'element' ) && element.getCustomProperty( 'dataPipeline:transparentRendering' ) ||
+		previous.is( 'element' ) && previous.getCustomProperty( 'dataPipeline:transparentRendering' )
+	) {
```
Comment (niegowski) [inline packages/ckeditor5-clipboard/src/utils/viewtoplaintext.ts:93]:
```suggestion
	if (
		element.is( 'element' ) && element.getCustomProperty( 'dataPipeline:transparentRendering' ) ||
		previous.is( 'element' ) && previous.getCustomProperty( 'dataPipeline:transparentRendering' )
	) {
```

**Comment by Dumluregn** [comment]
PR was changed due to the decision to move the integration points to the paste plain text and code block features.

File: packages/ckeditor5-clipboard/src/pasteplaintext.ts [hunk 3]
```diff
@@ -86,18 +81,22 @@ export default class PastePlainText extends Plugin {
 }
 
 /**
- * Returns true if specified `documentFragment` represents a plain text.
+ * Returns true if specified `documentFragment` represents the unformatted inline content.
```
Comment (scofalik) [inline packages/ckeditor5-clipboard/src/pasteplaintext.ts:84]:
```suggestion
 * Returns true if specified `documentFragment` represents an unformatted inline content.
```

File: packages/ckeditor5-clipboard/src/pasteplaintext.ts [hunk 4]
```diff
@@ -86,18 +81,22 @@ export default class PastePlainText extends Plugin {
...
+		if ( !model.schema.isInline( child ) ) {
+			return false;
+		}
 
-	const child = documentFragment.getChild( 0 )!;
+		const attributeKeys = Array.from( child.getAttributeKeys() );
 
-	if ( schema.isObject( child ) ) {
-		return false;
+		if ( attributeKeys.length !== 0 && attributeKeys.find( key => model.schema.getAttributeProperties( key ).isFormatting ) ) {
```
Comment (scofalik) [inline packages/ckeditor5-clipboard/src/pasteplaintext.ts:96]: Do we need this part `attributeKeys.length !== 0 &&` in this `if`?

File: packages/ckeditor5-code-block/src/codeblockediting.ts [hunk 5]
```diff
@@ -202,6 +203,31 @@ export default class CodeBlockEditing extends Plugin {
...
+				const selection = model.document.selection;
+
+				if ( !selection.anchor!.parent.is( 'element', 'codeBlock' ) ) {
+					return;
+				}
+
+				model.change( writer => {
+					const contentRange = writer.createRangeIn( data.content );
+
+					for ( const item of [ ...contentRange.getItems() ] ) {
```
Comment (scofalik) [inline packages/ckeditor5-code-block/src/codeblockediting.ts:221]: I think `for ( const item of contentRange.getItems() )` should be enough, no need to convert to array.

**Review by scofalik — CHANGES_REQUESTED** [review]
Very minor issues.

File: packages/ckeditor5-code-block/src/codeblockediting.ts (hunk 5 above)
Comment (Dumluregn) [reply packages/ckeditor5-code-block/src/codeblockediting.ts:221]: There's a removal in this loop. When it triggers, the iterator stops. If you have two merge fields in the pasted content, the second one will not be deleted.

**Pull request merged**